src/engine/sparqlExpressions/SparqlExpressionGenerators.h

```diff
@@ -200,6 +202,166 @@ CPP_template(typename Input, typename Transformation = ql::identity)(
+
+  SparseGeneratorRange(GeneratorRange generator, IndicesRange indices,
+                       size_t totalSize)
+      : generator_{std::move(generator)},
+        indices_{std::move(indices)},
+        totalSize_{totalSize} {}
+
+  class iterator {
+   private:
+    using GenRange = GeneratorRange;
+    using IdxRange = IndicesRange;
+
+    GenRange* generator_ = nullptr;
+    IdxRange* indices_ = nullptr;
+    ql::ranges::iterator_t<GenRange> genIt_{};
+    ql::ranges::sentinel_t<GenRange> genEnd_{};
+    ql::ranges::iterator_t<IdxRange> idxIt_{};
+    ql::ranges::sentinel_t<IdxRange> idxEnd_{};
+    size_t currentIndex_ = 0;
+    size_t totalSize_ = 0;
+
+    void skipToCurrentTarget() {
+      if (!generator_ || !indices_ || idxIt_ == idxEnd_) {
+        return;
+      }
+      auto target = *idxIt_;
+      AD_CONTRACT_CHECK(target < totalSize_);
+      while (currentIndex_ < target && genIt_ != genEnd_) {
+        ++genIt_;
+        ++currentIndex_;
+      }
+    }
+
+   public:
+    using iterator_category = std::input_iterator_tag;
+    using value_type = ql::ranges::range_value_t<GenRange>;
+    using difference_type = std::ptrdiff_t;
+    using reference = decltype(*genIt_);
+
+    iterator() = default;
+
+    iterator(GenRange* generator, IdxRange* indices, size_t totalSize)
+        : generator_{generator},
+          indices_{indices},
+          currentIndex_{0},
+          totalSize_{totalSize} {
+      if (generator_ && indices_) {
+        genIt_ = ql::ranges::begin(*generator_);
+        genEnd_ = ql::ranges::end(*generator_);
+        idxIt_ = ql::ranges::begin(*indices_);
+        idxEnd_ = ql::ranges::end(*indices_);
+        skipToCurrentTarget();
+      }
+    }
+
+    reference operator*() const { return *genIt_; }
+
+    iterator& operator++() {
+      if (!generator_ || idxIt_ == idxEnd_) {
+        return *this;
+      }
+      ++idxIt_;
+      if (idxIt_ == idxEnd_) {
+        return *this;
+      }
+      auto nextTarget = *idxIt_;
+      AD_CONTRACT_CHECK(nextTarget >= currentIndex_);
+      AD_CONTRACT_CHECK(nextTarget < totalSize_);
+      while (currentIndex_ < nextTarget && genIt_ != genEnd_) {
+        ++genIt_;
+        ++currentIndex_;
+      }
+      return *this;
+    }
+
+    void operator++(int) { (void)++(*this); }
+
+    friend bool operator==(const iterator& it, std::default_sentinel_t) {
+      return !it.generator_ || it.genIt_ == it.genEnd_ ||
+             it.idxIt_ == it.idxEnd_;
+    }
+
+    friend bool operator!=(const iterator& it, std::default_sentinel_t s) {
+      return !(it == s);
+    }
+  };
+
+  iterator begin() { return iterator{&generator_, &indices_, totalSize_}; }
+
+  std::default_sentinel_t end() { return {}; }
+};
+
+/// Return a generator that yields values only for the rows whose indices are
+/// contained in `indices`. The indices must be sorted in ascending order and
+/// must be strictly smaller than `numItems`.
+///
```

Comment:
I haven't looked at this in detail, 
but it 
1. looks reasonable in principle, but very complicated (not sure if it can be done simpler) or already exists
2. It only solves the problem halfway (if there is a subexpression that is complicated, that still will be evaluated).

I think for now, you can leave your old solution, and if desired then keep this for a follow-up PR if required.